### PR TITLE
EOS-15089: preliminary PR for CopyObject feature

### PR DESCRIPTION
**Pull request by evgeniy-brazhnikov**
Signed-off-by: evgeniy-brazhnikov <evgeniy.brazhnikov@seagate.com>

### CODE REVIEW

**Comment by evgeniy-brazhnikov** [comment]
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/140/

**Comment by evgeniy-brazhnikov** [comment]
jenkins-build.sh finished successfully on Dev VM too

File: server/s3_motr_writer.cc [hunk 1]
```diff
@@ -372,15 +372,28 @@ void S3MotrWiter::create_object_failed() {
...
-    std::shared_ptr<S3AsyncBufferOptContainer> buffer) {
+void S3MotrWiter::write_content(std::function<void(void)> on_success,
+                                std::function<void(void)> on_failed,
+                                S3BufferSequence buffer_sequence,
+                                size_t size_of_each_buf) {
   s3_log(S3_LOG_INFO, request_id, "Entering with layout_id = %d\n",
          layout_ids[0]);
 
+  assert(!buffer_sequence.empty());
+  assert(!(size_of_each_buf & 0xFFF));  // size_of_each_buf % 4096 == 0
```
Comment (knrajnambiar76) [inline server/s3_motr_writer.cc:383]: Will this be same as multiple of 4k ?

File: server/s3_put_chunk_upload_object_action.cc [hunk 2]
```diff
@@ -316,6 +316,14 @@ void S3PutChunkUploadObjectAction::validate_put_chunk_request() {
   s3_log(S3_LOG_DEBUG, nullptr, "Exiting\n");
 }
 
+void S3PutChunkUploadObjectAction::_set_layout_id(int layout_id) {
+  assert(layout_id > 0 && layout_id < 15);
```
Comment (knrajnambiar76) [inline server/s3_put_chunk_upload_object_action.cc:320]: layout id max is 9 right ?

File: server/s3_put_multiobject_action.cc [hunk 3]
```diff
@@ -310,6 +310,14 @@ void S3PutMultiObjectAction::fetch_firstpart_info_failed() {
   send_response_to_s3_client();
 }
 
+void S3PutMultiObjectAction::_set_layout_id(int layout_id) {
+  assert(layout_id > 0 && layout_id < 15);
```
Comment (knrajnambiar76) [inline server/s3_put_multiobject_action.cc:314]: max layout id is 9 I feel

File: server/s3_async_buffer_opt.cc [hunk 4]
```diff
@@ -134,28 +138,49 @@ std::pair<std::deque<evbuf_t*>, size_t> S3AsyncBufferOptContainer::get_buffers(
...
     }
-
-    evbuf_t* buf = NULL;
-    size_t len = 0;
     assert(ready_q.size() >= count_bufs_shared_for_read);
+
     for (size_t i = 0; i < count_bufs_shared_for_read; ++i) {
-      buf = ready_q.front();
-      processing_q.push_back(buf);
+      evbuf_t* p_ev_buf = ready_q.front();
```
[the remark this answers is not in the document]
Comment (evgeniy-brazhnikov) [reply server/s3_async_buffer_opt.cc:144]: It was a recommendation from C++'s author to make variables as local as possible.

File: server/s3_async_buffer_opt.cc [hunk 5]
```diff
@@ -18,19 +18,20 @@
...
       is_expecting_more(true),
-      count_bufs_shared_for_read(0) {
+      count_bufs_shared_for_read(0),
+      size_of_each_evbuf(size_of_each_buf) {
   s3_log(S3_LOG_DEBUG, "", "Constructor with size_of_each_buf = %zu\n",
          size_of_each_buf);
+
   // Should be multiple of 4k (Motr requirement)
-  assert(size_of_each_evbuf % 4096 == 0);
+  assert(!(size_of_each_evbuf & 0xFFF));  // size_of_each_evbuf % 4096 == 0
```
[the remark this answers is not in the document]
Comment (evgeniy-brazhnikov) [reply server/s3_async_buffer_opt.cc:34]: Divide operation is very expensive against bitwise ones.

File: server/s3_async_buffer_opt.cc [hunk 6]
```diff
@@ -108,18 +111,19 @@ bool S3AsyncBufferOptContainer::add_content(evbuf_t* buf, bool is_first_buf,
 // after get_buffers call.
 // expected_content_size should be multiple of libevent read mempool item size
 // except for the last payload
-std::pair<std::deque<evbuf_t*>, size_t> S3AsyncBufferOptContainer::get_buffers(
+S3BufferSequence S3AsyncBufferOptContainer::get_buffers(
```
[the remark this answers is not in the document]
Comment (evgeniy-brazhnikov) [reply server/s3_async_buffer_opt.cc:114]: It's the major purpose of this PR.

File: server/s3_async_buffer_opt.h [hunk 7]
```diff
@@ -55,15 +55,17 @@ class S3AsyncBufferOptContainer {
   size_t count_bufs_shared_for_read;
 
  public:
-  S3AsyncBufferOptContainer(size_t size_of_each_buf);
+  const size_t size_of_each_evbuf;  // ideally 4k/8k/16k
```
[the remark this answers is not in the document]
Comment (evgeniy-brazhnikov) [reply server/s3_async_buffer_opt.h:58]: Not sure. It's the size of libevent buffers.

File: server/s3_motr_writer.cc (hunk 1 above)
Comment (evgeniy-brazhnikov) [reply server/s3_motr_writer.cc:383]: In this case - exactly

File: server/s3_put_chunk_upload_object_action.cc (hunk 2 above)
Comment (evgeniy-brazhnikov) [reply server/s3_put_chunk_upload_object_action.cc:320]: 14 in s3config

File: server/s3_put_multiobject_action.cc (hunk 3 above)
Comment (evgeniy-brazhnikov) [reply server/s3_put_multiobject_action.cc:314]: I inserted these asserts for something like uninitialized variable or other rude errors.

File: server/s3_async_buffer_opt.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (knrajnambiar76) [reply server/s3_async_buffer_opt.h:58]: For write we have 16Kbuffer
For read also now 16K buffers ?

File: server/s3_put_chunk_upload_object_action.cc (hunk 2 above)
Comment (knrajnambiar76) [reply server/s3_put_chunk_upload_object_action.cc:320]: ok
In s3_obj_layout_mapping.yaml we have max till 9. I feel currently max is 9 and can go up later.

File: server/s3_async_buffer_opt.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (evgeniy-brazhnikov) [reply server/s3_async_buffer_opt.h:58]: No. S3_MOTR_UNIT_SIZE * S3_MOTR_MAX_UNITS_PER_REQUEST

**Review by knrajnambiar76 — APPROVED** [review]

**Pull request closed without merging**